vizro-core/tests/unit/vizro/models/_action/test_action.py

```diff
@@ -282,67 +271,104 @@ def test_action_callback_function_return_value_valid(
         result = action._action_callback_function(inputs={}, outputs=callback_outputs)
         assert result == expected_function_return_value
 
+    @pytest.mark.parametrize("callback_outputs", [[], {}, None])
+    @pytest.mark.parametrize(
+        "custom_action_function_mock_return",
+        [False, 0, "", [], (), {}],
+        indirect=True,
+    )
+    def test_action_callback_function_no_outputs_return_value_exists(
+        self, custom_action_function_mock_return, callback_outputs
+    ):
+        action = Action(function=custom_action_function_mock_return())
+        with pytest.raises(
+            ValueError,
+            match="Action function has returned a value but the action has no defined outputs.",
+        ):
+            action._action_callback_function(inputs={}, outputs=callback_outputs)
+
     @pytest.mark.parametrize(
         "custom_action_function_mock_return, callback_outputs",
         [
-            (None, ["component_1_property", "component_2_property"]),
-            (False, []),
-            (0, []),
-            (123, []),
-            (123, ["component_1_property", "component_2_property"]),
-            ("", []),
-            ("ab", []),
-            ("ab", ["component_1_property", "component_2_property", "component_3_property"]),
-            ((), []),
-            (("new_value"), []),
-            (("new_value"), ["component_1_property", "component_2_property"]),
-            (("new_value", "new_value_2"), []),
-            (("new_value", "new_value_2"), ["component_1_property", "component_2_property", "component_3_property"]),
-            ([], []),
-            (["new_value"], []),
-            (["new_value"], ["component_1_property", "component_2_property"]),
-            (["new_value", "new_value_2"], []),
-            (["new_value", "new_value_2"], ["component_1_property", "component_2_property", "component_3_property"]),
-            ({}, []),
-            ({"component_1_property": "new_value"}, []),
-            ({"component_1_property": "new_value"}, ["component_1_property", "component_2_property"]),
-            ({"component_1_property": "new_value", "component_2_property": "new_value_2"}, []),
-            (
-                {"component_1_property": "new_value", "component_2_property": "new_value_2"},
-                ["component_1_property", "component_2_property", "component_3_property"],
-            ),
+            (None, ["component_1_property"]),
+            (False, ["component_1_property"]),
+            (0, ["component_1_property"]),
+            (123, ["component_1_property"]),
```

Comment:
Please could we add a case with `custom_action_function_mock_return` a dictionary here?

Reply:
What do you mean? The following case will pass: `({'a': 1}, ["component_1_property"])` (and it should pass). Remember that we transform outputs to outputs[0] before `_action_callback_function` happens